### PR TITLE
feat: puppet facts

### PR DESCRIPTION
**Pull request by bdemetris**
Uses puppet facts command into an interface, and flattens the map so sal will accept k,v pairs.

### CODE REVIEW

**Comment by clburlison** [comment]
As a bonus, this implementation works with Chef's Ohai. I'm sure you're aware of it but test cases are going to only become more valuable as this project evolves.

My other feedback would be to rename `puppet.go` to `factor.go`. That change would allow a `GetGrainsFacts` & `GetOhaiFacts` to be added quite easily.

**Comment by bdemetris** [comment]
@clburlison yep i agree, went ahead and renamed `puppet.go`

we're in a bit of a time crunch to get this out.  i will open a few issues that are going to bubble up as a result of some things i rushed.

**Comment by grahamgilbert** [comment]
The code obvs works, but not sure on the CI failure. Looks like a transient failure as travis couldn't pull all (any?) of the things it needs.

File: reports/emulator_base64bz2.go [hunk 1]
```diff
@@ -3,23 +3,58 @@ package reports
...
+	// TODO clean this up, loadconfig should do this work
+	dir, err := filepath.Abs(filepath.Dir(os.Args[0]))
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	s := filepath.Join(dir, "config.json")
+	conf, err := loadConfig(s)
+	if err != nil {
+		log.Fatal(err)
```
Comment (groob) [inline reports/emulator_base64bz2.go]: turn this into `return errors.Wrap(err, "loading config json")`

File: reports/emulator_base64bz2.go [hunk 2]
```diff
@@ -3,23 +3,58 @@ package reports
...
+	}
+
+	// switch for whats "facts" we would send sal
+	var facts map[string]interface{}
+	switch conf.Management.Tool {
+	case "puppet":
+		facts, err = GetPuppetFacts()
+		fmt.Println("got here")
+		if err != nil {
+			errors.Wrap(err, "puppet switch: failed to get facts")
```
Comment (groob) [inline reports/emulator_base64bz2.go]: Add a `return` in front of `errors.Wrap`, otherwise it's a noop.

File: reports/emulator_base64bz2.go [hunk 3]
```diff
@@ -3,23 +3,58 @@ package reports
...
 }
 
 // BuildBase64bz2Report will return a compressed and encoded string of our report struct
 func BuildBase64bz2Report() (string, error) {
 
+	// gets the absolute path to the config file
+	// TODO clean this up, loadconfig should do this work
+	dir, err := filepath.Abs(filepath.Dir(os.Args[0]))
+	if err != nil {
+		log.Fatal(err)
```
Comment (groob) [inline reports/emulator_base64bz2.go]: replace `log.Fatal` with `return errors.Wrap(err, "reason it failed")`

File: reports/facter.go [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+	}
+
+	v := pf["values"].(map[string]interface{})
+
+	pf = Flatten(v)
+
+	return pf, nil
+}
+
+// Flatten takes a map[string]interface and flattens it
```
Comment (groob) [inline reports/facter.go]: What are you flattening the results to? Are they always strings?

Comment (bdemetris) [reply reports/facter.go]: well.  in this case they are all strings :) but would like to reflect type.

**Pull request merged**